lib/balanced_vrp_clustering.rb

```diff
@@ -243,34 +243,44 @@ def recompute_centroids
           end
         }
 
-        swap_centroids_with_capacity_problems
+        swap_a_centroid_with_capacity_problems
 
         @iterations += 1
       end
 
-      def swap_centroids_with_capacity_problems
+      def swap_a_centroid_with_capacity_problems
         @clusters_with_capacity_violation.map.with_index.sort_by{ |arr, _i| -arr.size }.each{ |preferred_clusters, violated_cluster|
           break if preferred_clusters.empty?
 
-          favorite_cluster = if rand < 0.10
-                               # TODO: elimination/determination of units per cluster shoudl be done at the beginning
-                               # Each centroid should know what matters to them.
-                               the_units_that_matter = @centroids[violated_cluster][3].select{ |i, v| i && v.positive? }.keys & @strict_limitations[violated_cluster].select{ |i, v| i && v}.keys
+          # TODO: elimination/determination of units per cluster shoudl be done at the beginning
+          # Each centroid should know what matters to them.
+          the_units_that_matter = @centroids[violated_cluster][3].select{ |i, v| i && v.positive? }.keys & @strict_limitations[violated_cluster].select{ |i, v| i && v}.keys
+
+          # TODO: only consider the clusters that are "compatible" with this cluster -- i.e., they can serve the points of this cluster and vice-versa
+          favorite_clusters = @centroids.map.with_index.select{ |c, i|
+            the_units_that_matter.all?{ |unit| # cluster to be swapped should
+              c[3][unit] < 0.99 * @strict_limitations[violated_cluster][unit] && # be less loaded
+                @strict_limitations[i][unit] >= 1.01 * @centroids[violated_cluster][3][unit] && # have more capacity
+                @clusters[violated_cluster].data_items.all?{ |d_i| @compatibility_function.call(d_i, @centroids[i]) } &&
+                @clusters[i].data_items.all?{ |d_i| @compatibility_function.call(d_i, @centroids[violated_cluster]) }
+            } # and they should be able to serve eachothers's points
+          }
 
-                               @centroids.map.with_index.select{ |c, _i|
-                                 the_units_that_matter.all?{ |unit| c[3][unit] < @centroids[violated_cluster][3][unit] }
-                               }.min_by{ |c, i|
-                                 the_units_that_matter.sum{ |unit| c[3][unit].to_f / @strict_limitations[i][unit] * (rand(0.90) + 0.1) } # give others some chance with randomness
-                               }[1]
-                             else
-                               preferred_clusters.uniq.sample
-                             end
+          if favorite_clusters.empty?
+            # puts "cannot swap #{violated_cluster + 1} because no one can save it"
```

Comment:
```suggestion
```

Reply:
or transform into comment ?

Reply:
It is left as a comment in purpose. Because it is the initial dev phase, we need it regularly but we don't want the log of the mt-web is filled with it.